magento/pwa-studio
# [featbug]: Implement UX around zero total checkout

## Description

Through a multitude of use cases you may find yourself with a $0 cart total. This PR handles those cases but also the inverse - going from a $0 cart to some total greater than zero. It also handles the case where a user has refreshed (or left the checkout page) after going to the review step.

## Related Issue





Closes [PWA-563](https://jira.corp.magento.com/browse/PWA-563).

## Acceptance 

### Verification Steps

To verify this PR you need to check **all** the ways a user may interact with a cart and a $0 total.

* Check how the Payment Information component behaves when the total changes within `/checkout` and also if the total changes outside of checkout. 
* Check how it behaves after having already had some info such as a CC that was submitted (click through to "review order" but then go back to `/cart` and remove the GC).

Basically any way that the cart could get to $0 or become > $0 needs to be checked. 

In development I found the best way to mimic this was to use the gift card component to add/remove a gift card that covered the balance. You can then add/remove that card on the /cart page or the /checkout page and navigate back and forth to see how the checkout view responds.

### Use Cases

I've tried to enumerate the use cases I had to work through for this code. I hope I covered them all.

* Zero Total Checkout must be enabled.
* "Adding a gift card" in this case is effectively used to make the total $0. You could accomplish the same with a coupon, or by having two items, a gift card with a value that covers one item and then remove another item that causes the total to become $0.
* We avoid making unnecessary gql calls. If a valid method is already selected on cart then we don't need to re-submit data when clicking "review order".

**User gets to "payment info" and total is $0**

1) On /cart, add a gift card which causes total to be $0. 
2) Go to /checkout and enter shipping info and proceed.
3) Select shipping method and proceed. You should see "free" summary view.
4) Click "review order". You should see review order view.
5) Refresh the page. You should still see "free" summary view.
6) Delete `apollo-cache-persist` and refresh the page. You should see the "free" summary view.
7) Click "review order" and "place order". It should work.

**When "free" is selected but cart total changes on checkout**

1) On /cart, add a gift card which causes total to be $0.
2) Go to /checkout and enter shipping info and proceed.
3) Select shipping method and proceed. You should see "free" summary view.
4) Remove the gift card. The braintree form should render.

**When "free" is selected but cart total changes outside of checkout**

1) On /cart, add a gift card which causes total to be $0.
2) Go to /checkout and enter shipping info and proceed.
3) Select shipping method and proceed. You should see "free" summary view.
4) Go back to /cart and remove the gift card.
5) Go /checkout. The braintree form should render.

**User submits a CC but then adds a gift card**

1) Go through /checkout with a non-zero total.
2) Enter CC data and click "review order". You should see the CC summary view.
3) Refresh the page. The CC summary view is still visible.
4) Add a gift card. The "free" summary should be visible.

### Verify existing functionality

**User submits a CC and places order**
1) Go through /checkout with a non-zero total.
2) Enter CC data and click "review order". You should see the CC summary view.
3) Refresh the page. The CC entry form is visible. Enter CC data again.
4) Click "review" and "place order". It should work.

## Screenshots / Screen Captures (if appropriate)

[![Image from Gyazo](https://i.gyazo.com/8a3182bee7dac5adc5b4a4e39260b924.gif)](https://gyazo.com/8a3182bee7dac5adc5b4a4e39260b924)

## Checklist

* I have updated the documentation accordingly, if necessary.
* I have added tests to cover my changes, if necessary.
